# QUESTION

## 1. Introduction
<p><img src="https://assets.datacamp.com/production/project_1197/img/google_play_store.png" alt="Google Play logo"></p>
<p>Mobile apps are everywhere. They are easy to create and can be very lucrative from the business standpoint. Specifically, Android is expanding as an operating system and has captured more than 74% of the total market<sup><a href="https://www.statista.com/statistics/272698/global-market-share-held-by-mobile-operating-systems-since-2009">[1]</a></sup>. </p>
<p>The Google Play Store apps data has enormous potential to facilitate data-driven decisions and insights for businesses. In this notebook, we will analyze the Android app market by comparing ~10k apps in Google Play across different categories. We will also use the user reviews to draw a qualitative comparision between the apps.</p>
<p>The dataset you will use here was scraped from Google Play Store in September 2018 and was published on <a href="https://www.kaggle.com/lava18/google-play-store-apps">Kaggle</a>. Here are the details: <br>
<br></p>
<div style="background-color: #efebe4; color: #05192d; text-align:left; vertical-align: middle; padding: 15px 25px 15px 25px; line-height: 1.6;">
    <div style="font-size:20px"><b>datasets/apps.csv</b></div>
This file contains all the details of the apps on Google Play. There are 9 features that describe a given app.
<ul>
    <li><b>App:</b> Name of the app</li>
    <li><b>Category:</b> Category of the app. Some examples are: ART_AND_DESIGN, FINANCE, COMICS, BEAUTY etc.</li>
    <li><b>Rating:</b> The current average rating (out of 5) of the app on Google Play</li>
    <li><b>Reviews:</b> Number of user reviews given on the app</li>
    <li><b>Size:</b> Size of the app in MB (megabytes)</li>
    <li><b>Installs:</b> Number of times the app was downloaded from Google Play</li>
    <li><b>Type:</b> Whether the app is paid or free</li>
    <li><b>Price:</b> Price of the app in US$</li>
    <li><b>Last Updated:</b> Date on which the app was last updated on Google Play </li>

</ul>
</div>
<div style="background-color: #efebe4; color: #05192d; text-align:left; vertical-align: middle; padding: 15px 25px 15px 25px; line-height: 1.6;">
    <div style="font-size:20px"><b>datasets/user_reviews.csv</b></div>
This file contains a random sample of 100 <i>[most helpful first](https://www.androidpolice.com/2019/01/21/google-play-stores-redesigned-ratings-and-reviews-section-lets-you-easily-filter-by-star-rating/)</i> user reviews for each app. The text in each review has been pre-processed and passed through a sentiment analyzer.
<ul>
    <li><b>App:</b> Name of the app on which the user review was provided. Matches the `App` column of the `apps.csv` file</li>
    <li><b>Review:</b> The pre-processed user review text</li>
    <li><b>Sentiment Category:</b> Sentiment category of the user review - Positive, Negative or Neutral</li>
    <li><b>Sentiment Score:</b> Sentiment score of the user review. It lies between [-1,1]. A higher score denotes a more positive sentiment.</li>

</ul>
</div>
<p>From here on, it will be your task to explore and manipulate the data until you are able to answer the three questions described in the instructions panel.<br></p>

# **CODE BEGINS HERE**

## TASK 1

First, let's import the relevant modules and load the two csv files as a pandas DataFrame.

In [1]:
import pandas as pd

In [2]:
apps = pd.read_csv('datasets/apps.csv')
reviews = pd.read_csv('datasets/user_reviews.csv')

Let's get a thorough insight on the dataset using the .info() and .head() methods on the two DataFrames.

In [3]:
apps.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9659 entries, 0 to 9658
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   App           9659 non-null   object 
 1   Category      9659 non-null   object 
 2   Rating        8196 non-null   float64
 3   Reviews       9659 non-null   int64  
 4   Size          8432 non-null   float64
 5   Installs      9659 non-null   object 
 6   Type          9659 non-null   object 
 7   Price         9659 non-null   float64
 8   Last Updated  9659 non-null   object 
dtypes: float64(3), int64(1), object(5)
memory usage: 679.3+ KB


In [4]:
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 64295 entries, 0 to 64294
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   App                 64295 non-null  object 
 1   Review              37427 non-null  object 
 2   Sentiment Category  37432 non-null  object 
 3   Sentiment Score     37432 non-null  float64
dtypes: float64(1), object(3)
memory usage: 2.0+ MB


In [5]:
reviews.head()

App                                             Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   

  Sentiment Category  Sentiment Score  
0           Positive             1.00  
1           Positive             0.25  
2                NaN              NaN  
3           Positive             0.40  
4           Positive             1.00

The .sample() method is used to generate a random sample of data from the DataFrame. It is better than the .head() method as a more diverse sample of data can be examined.

In [6]:
apps.sample(10)

App             Category  Rating  \
782         ColorFul - Adult Coloring Book        ENTERTAINMENT     4.6   
5894  Fantasy theme dark bw black building       ART_AND_DESIGN     4.8   
993              Wendy’s – Food and Offers       FOOD_AND_DRINK     3.4   
8078                     EB Remote Deposit             BUSINESS     NaN   
7912                             DX Alerts            LIFESTYLE     4.5   
8812                 Power Plug EV charger  MAPS_AND_NAVIGATION     3.7   
4579                         AR Remote Car               FAMILY     3.2   
2334                        beIN SPORTS TR               SPORTS     4.0   
6003                               GITZ.bz             SHOPPING     4.0   
1804              palmPEDi: Pediatric Tape              MEDICAL     4.6   

      Reviews  Size     Installs  Type  Price       Last Updated  
782     50725  19.0   5,000,000+  Free   0.00       May 25, 2018  
5894       41   1.9       5,000+  Free   0.00     March 20, 2018  
993      6507  18.0   1,000,000+  Free   0.00     April 12, 2018  
8078        0   9.0          10+  Free   0.00       May 10, 2018  
7912       22  23.0       1,000+  Free   0.00      July 19, 2016  
8812       13   3.9       1,000+  Free   0.00  December 22, 2014  
4579     1354  30.0     100,000+  Free   0.00    August 12, 2017  
2334   135763  13.0  10,000,000+  Free   0.00     August 2, 2018  
6003        4   2.4          10+  Free   0.00     August 4, 2017  
1804       66   2.6       5,000+  Paid   0.99   December 2, 2013

The first task was to convert the values on 'Installs' column in the dataset to an integer value. Originally, the 'Installs' column contained String values.


To convert the column to integer, we first need to remove special signs, like the '+' and ',' signs present in values on the 'Installs' column. We did that using the replace() method on the Series. The result was a string, but without the operator signs. We then used the to_numeric() function from Pandas to convert the Series from string (referred to in pandas as Object data type) to int64 type.

In [7]:
# Removing special characters from the 'Installs' column.

chars_to_replace = [',', '+']

for char in chars_to_replace:
    apps['Installs'] = apps['Installs'].str.replace(char, '', regex=False)
apps['Installs'] = pd.to_numeric(apps['Installs'], errors='raise')

We then use the .info() method on the 'Installs' column to see that the Installs column is now of data type int64.

In [8]:
apps['Installs'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 9659 entries, 0 to 9658
Series name: Installs
Non-Null Count  Dtype
--------------  -----
9659 non-null   int64
dtypes: int64(1)
memory usage: 75.6 KB


Again, let's have a look at the modified apps DataFrame.

In [9]:
apps.sample(10)

App          Category  Rating  Reviews  \
5612                     BQ Llicència            SPORTS     NaN       18   
6334                          IQ Test      PRODUCTIVITY     4.2     1623   
4231  ALPHA - Artificial Intelligence            FAMILY     4.0       51   
4718               AV HD Video Player            FAMILY     3.8        6   
6265                     Bhunaksha CG             TOOLS     4.1       67   
8566      The Great Wobo Escape Ep. 1              GAME     4.2     5449   
3508     Car Parking Crane N Drifting            FAMILY     4.2      106   
2767           Pushbullet - SMS on PC      PRODUCTIVITY     4.5   176873   
7995               My Ooredoo Algeria             TOOLS     4.2     3606   
2428      KakaoMap - Map / Navigation  TRAVEL_AND_LOCAL     4.1    76779   

      Size  Installs  Type  Price        Last Updated  
5612   NaN      5000  Free    0.0    January 30, 2016  
6334   1.8    100000  Free    0.0       March 6, 2018  
4231   3.2      1000  Free    0.0        June 3, 2018  
4718  23.0      1000  Free    0.0    January 19, 2018  
6265   4.1     10000  Free    0.0         May 7, 2018  
8566  20.0    500000  Free    0.0  September 13, 2016  
3508  71.0     50000  Free    0.0   February 26, 2018  
2767   4.5   1000000  Free    0.0    October 30, 2017  
7995  17.0    100000  Free    0.0       June 24, 2018  
2428   NaN  10000000  Free    0.0       July 30, 2018

We can also see that the Date is of type 'str'. There is a special data type in Python, called datetime, which is used to process date/ time values. So, let's look at converting the values from the 'Last Updated' column to type datetime.

pandas has a valuable function called to_datetime(), which is similar to the to_numeric() function we used earlier. However, in this function, we need to specify the format of the date/ time in the present column. For this, we can use metacharacters. Metacharacters are special characters that represent certain values in Python.

I looked up the key for metacharacters at the strftime docs available at https://docs.python.org/3/library/datetime.html#strftime-and-strptime-behavior.

In [10]:
date = pd.to_datetime(apps['Last Updated'], format = '%B %d, %Y')
apps['Last Updated'] = date

In [11]:
apps['Last Updated'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 9659 entries, 0 to 9658
Series name: Last Updated
Non-Null Count  Dtype         
--------------  -----         
9659 non-null   datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 75.6 KB


## END OF TASK 1

***

# TASK 2

For the next specified task, we were to find the number of apps in each category, the average price, and the average rating. The answer was to be saved as a DataFrame called app_category_info, with the four columns named: Category, Number of apps, Average price, Average rating.

Now, we need to group the rows by 'Category' as specified in the question. Then, we count the number of apps using the count() method on the GroupBy object generated using the groupby() method. Similarly, we use the agg() method to calculate the averages for the price and rating for all groups in the category. 

In [12]:
no_apps_df = apps.groupby('Category')['App'].count()
avg_price_df = apps.groupby('Category')['Price'].agg('mean')
avg_rating_df = apps.groupby('Category')['Rating'].agg('mean')

In [13]:
app_category_info = pd.DataFrame(data = {
                     'Number of apps':no_apps_df, 
                     'Average price':avg_price_df, 
                     'Average rating':avg_rating_df})

In [14]:
# DATAFRAME CREATED, #2 DONE

app_category_info.reset_index()

Category  Number of apps  Average price  Average rating
0        ART_AND_DESIGN              64       0.093281        4.357377
1     AUTO_AND_VEHICLES              85       0.158471        4.190411
2                BEAUTY              53       0.000000        4.278571
3   BOOKS_AND_REFERENCE             222       0.539505        4.344970
4              BUSINESS             420       0.417357        4.098479
5                COMICS              56       0.000000        4.181481
6         COMMUNICATION             315       0.263937        4.121484
7                DATING             171       0.160468        3.970149
8             EDUCATION             119       0.150924        4.364407
9         ENTERTAINMENT             102       0.078235        4.135294
10               EVENTS              64       1.718594        4.435556
11               FAMILY            1832       1.309967        4.179664
12              FINANCE             345       8.408203        4.115563
13       FOOD_AND_DRINK             112       0.075714        4.172340
14                 GAME             959       0.296465        4.247368
15   HEALTH_AND_FITNESS             288       0.223438        4.243033
16       HOUSE_AND_HOME              74       0.000000        4.150000
17   LIBRARIES_AND_DEMO              84       0.011786        4.178125
18            LIFESTYLE             369       6.398022        4.093355
19  MAPS_AND_NAVIGATION             131       0.205725        4.036441
20              MEDICAL             395       2.520759        4.166552
21   NEWS_AND_MAGAZINES             254       0.015669        4.121569
22            PARENTING              60       0.159667        4.300000
23      PERSONALIZATION             376       0.400213        4.332215
24          PHOTOGRAPHY             281       0.420925        4.157414
25         PRODUCTIVITY             374       0.670936        4.183389
26             SHOPPING             202       0.027129        4.230000
27               SOCIAL             239       0.066820        4.247291
28               SPORTS             325       0.307692        4.216154
29                TOOLS             827       0.323156        4.039554
30     TRAVEL_AND_LOCAL             219       0.228082        4.069519
31        VIDEO_PLAYERS             163       0.064172        4.044595
32              WEATHER              79       0.410380        4.243056

## END OF TASK 2

***

# TASK 3

For the third task, we were to find the top 10 free FINANCE apps having the highest average sentiment score. A DataFrame named top_10_user_feedback with the columns App and Sentiment Score was to be created, with the average sentiment scores sorted in descending order.

We can see from the above output that there are a total of 345 apps in the 'Finance' category.

Now, to find the average sentiment scores for each app, we need to first have a dataframe consisting of all the apps, their categories and the Sentiment Scores. This means we need to access values from both apps and reviews DataFrames. This can be done by joining the two dataframes on the 'App' column.

In [15]:
score_apps_table = apps.merge(reviews, on='App', how='inner')

In [16]:
score_apps_table.sample(5)

App            Category  Rating  Reviews  Size  \
20403    Binaural Beats Therapy  HEALTH_AND_FITNESS     4.3    13724   NaN   
34483  Dr. Panda Town: Vacation              FAMILY     3.8    10366  78.0   
5351                 Chrome Dev       COMMUNICATION     4.4    63543   NaN   
32659  Bus Rush: Subway Edition                GAME     4.1   753043  57.0   
52078     Beautiful Widgets Pro     PERSONALIZATION     4.2    97890  14.0   

        Installs  Type  Price Last Updated  \
20403    1000000  Free   0.00   2017-08-17   
34483    1000000  Free   0.00   2018-07-24   
5351     5000000  Free   0.00   2018-08-02   
32659  100000000  Free   0.00   2018-08-01   
52078    1000000  Paid   2.49   2016-08-24   

                                                  Review Sentiment Category  \
20403  Changed rating 3 star. The UI scrolling lags r...           Positive   
34483  Why everyone kept saying needs money? I got fr...           Positive   
5351   Can't Add Extensions. Battery Hog. Homepage No...            Neutral   
32659                                                NaN                NaN   
52078  I've used BW several years always enjoyed it. ...           Positive   

       Sentiment Score  
20403         0.033333  
34483         0.125000  
5351          0.000000  
32659              NaN  
52078         0.028283

Let's now create a DataFrame that contains only the free apps from the FINANCE category, as required by the question.

In [17]:
finance_scores_table = score_apps_table[(score_apps_table['Category'] == 'FINANCE') & (score_apps_table['Type'] == 'Free')]

The DataFrame contains 2200 columns, even though we found out above that there are only 328 apps in the 'Finance' category. This is because the same app contains multiple reviews.

In [18]:
top_10_user_feedback = pd.DataFrame(finance_scores_table.groupby('App')['Sentiment Score'].agg('mean'))

In [19]:
top_10_user_feedback.groupby('App')['Sentiment Score'].agg('mean').sort_values(ascending = False)

App
BBVA Spain                                         0.515086
Associated Credit Union Mobile                     0.388093
BankMobile Vibe App                                0.353455
A+ Mobile                                          0.329592
Current debit card and app made for teens          0.327258
BZWBK24 mobile                                     0.326883
Even - organize your money, get paid early         0.283929
Credit Karma                                       0.270052
Fortune City - A Finance App                       0.266966
Branch                                             0.264230
ACE Elite                                          0.252171
Digit Save Money Automatically                     0.250246
CNBC: Breaking Business News & Live Market Data    0.247704
CreditWise from Capital One                        0.239454
Expense IQ Money Manager                           0.227506
BBVA Compass Banking                               0.205590
Bank of America Mobile Banking      

In [20]:
top_10_user_feedback = pd.DataFrame(top_10_user_feedback.groupby('App')['Sentiment Score'].agg('mean').sort_values(ascending=False)[:10])

In [21]:
top_10_user_feedback

Sentiment Score
App                                                        
BBVA Spain                                         0.515086
Associated Credit Union Mobile                     0.388093
BankMobile Vibe App                                0.353455
A+ Mobile                                          0.329592
Current debit card and app made for teens          0.327258
BZWBK24 mobile                                     0.326883
Even - organize your money, get paid early         0.283929
Credit Karma                                       0.270052
Fortune City - A Finance App                       0.266966
Branch                                             0.264230

From this result, we can see that the app `BBVA Spain` was the best free app on the Finance category on Google Play Store.

## END OF TASK 3

***